oauth/client_factory.go

```diff
@@ -14,156 +14,139 @@
-		return nil, fmt.Errorf("failed to add authz to the scheme: %w", err)
+	resp, err := client.Do(req)
+	if err != nil {
+		lg.Error(err, "failed to request the workspace API", "url", wsEndpoint)
+		return "", fmt.Errorf("error performing HTTP request for workspace context to %s: %w", wsEndpoint, err)
 	}
 
-	return options, nil
-}
-
-// baseKubernetesConfig returns proper configuration based on given cli args. `userAuthentication=true` allows setting apiserver url in cli args
-// and enables using bearer token in authorization header to authenticate kubernetes requests. This is used for requests we're doing on behalf of end-user.
-// If `kubeconfig` is set in cli args, it is always used.
-func baseKubernetesConfig(args *ClientFactoryConfig, userAuthentication bool) (*rest.Config, error) {
-	if args.KubeConfig != "" {
-		cfg, err := clientcmd.BuildConfigFromFlags("", args.KubeConfig)
-		if err != nil {
-			return nil, fmt.Errorf("failed to create rest configuration: %w", err)
-		}
-
-		return cfg, nil
-	} else if userAuthentication && args.ApiServer != "" {
-		// here we're essentially replicating what is done in rest.InClusterConfig() but we're using our own
-		// configuration - this is to support going through an alternative API server to the one we're running with...
-		// Note that we're NOT adding the Token or the TokenFile to the configuration here. This is supposed to be
-		// handled on per-request basis...
-		cfg := &rest.Config{}
-
-		apiServerUrl, err := url.Parse(args.ApiServer)
-		if err != nil {
-			return nil, fmt.Errorf("failed to parse the API server URL: %w", err)
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			lg.Error(err, "Failed to close response body doing workspace fetch")
 		}
+	}()
 
-		cfg.Host = "https://" + apiServerUrl.Host
+	if resp.StatusCode != http.StatusOK {
+		lg.Info("unexpected return code for workspace api", "url", wsEndpoint, "code", resp.StatusCode)
+		return "", fmt.Errorf("bad status (%d) when performing HTTP request for workspace context to %v: %w", resp.StatusCode, wsEndpoint, err)
+	}
 
-		tlsConfig := rest.TLSClientConfig{}
+	bodyBytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		lg.Error(err, "failed to read the workspace API response", "error", err.Error())
+		return "", fmt.Errorf("failed to read the workspace API response: %w", err)
+	}
 
-		if args.ApiServerCAPath != "" {
-			// rest.InClusterConfig is doing this most possibly only for early error handling so let's do the same
-			if _, err := certutil.NewPool(args.ApiServerCAPath); err != nil {
-				return nil, fmt.Errorf("expected to load root CA config from %s, but got err: %w", args.ApiServerCAPath, err)
-			} else {
-				tlsConfig.CAFile = args.ApiServerCAPath
+	wsList := &v1alpha1.WorkspaceList{}
+	if err := json.Unmarshal(bodyBytes, wsList); err != nil {
+		return "", fmt.Errorf("%w '%s': %s", errUnableToParseWorkspaceResponse, namespace, err.Error())
+	}
+	for _, ws := range wsList.Items {
```

Comment:
what about moving all the fetching of workspaces to own function and do the finding after? It's a a bit lost here with all the mess around fetching, parsing etc. Sth like:
```
workspaces, err := fetchWorkspaces(...)
if err != nil ...

for _, ws := range workspaces {
...
}
```